### PR TITLE
p2p: support protocol version negotiation

### PR DESCRIPTION
**Pull request by karalabe**

### CODE REVIEW

File: p2p/peer.go [hunk 1]
```diff
@@ -249,15 +249,22 @@ func countMatchingProtocols(protocols []Protocol, caps []Cap) int {
...
+	sort.Sort(capsByNameAndVersion(caps))
 	offset := baseProtocolLength
 	result := make(map[string]*protoRW)
+
 outer:
 	for _, cap := range caps {
 		for _, proto := range protocols {
-			if proto.Name == cap.Name && proto.Version == cap.Version && result[cap.Name] == nil {
+			if proto.Name == cap.Name && proto.Version == cap.Version {
+				// If an old protocol version matched, revert it
```
Comment (fjl) [inline p2p/peer.go]: This assumes that `protocols` is sorted by version. 
Idea:
```go
if old := result[cap.Name]; old != nil && old.Version < proto.Version {
```

File: p2p/peer_test.go [hunk 2]
```diff
@@ -196,3 +196,98 @@ func TestNewPeer(t *testing.T) {
 
 	p.Disconnect(DiscAlreadyConnected) // Should not hang
 }
+
+func TestMatchProtocols(t *testing.T) {
+	tests := []struct {
+		Local  []Cap
+		Remote []Protocol
+		Match  map[string]protoRW
```
Comment (fjl) [inline p2p/peer_test.go]: This is flipped: the remote end presents caps, protocols are local.

File: p2p/peer.go (hunk 1 above)
Comment (karalabe) [inline p2p/peer.go]: It shouldn't assume that. Caps is sorted by name, then version. So a smaller version will always match faster than a larger version due to caps being in the outer loop.

File: p2p/peer_test.go (hunk 2 above)
Comment (karalabe) [inline p2p/peer_test.go]: Whoops, will fix.

File: p2p/peer.go (hunk 1 above)
Comment (fjl) [inline p2p/peer.go]: I see it now. There is also a test for it.

**Comment by karalabe** [comment]
PTAL

**Comment by obscuren** [comment]
:+1:

**Pull request merged**